### ISSUE TITLE
Incorrect usage of StagingArea.

### ISSUE THREAD

**Issue by TimZaman** [avolkov1/keras_experiments#2]
Hey Alex!

Awesome that you're using the StagingArea. However, the usage of it is incorrect, and you won't see any speedup.

You're correctly aggregating a list of the staging area ops (you call enqueue ops). Those enqueue ops you'll eventually pass to the `run` function in your monkey-patched back-end.
What's happening now is, is that you're not pre-queueing anything. That means that if you run:
`run([enqueue_ops, your_model])`
Is that `your_model` will try to fetch from the staging_area, which is empty, and it will block. It will block, until your `enqueue_ops` is actually done `put`ting things in the StagingArea. Once something is in there, `your_model` will be unblocked, and `get` from your StagingArea. You see, this is no different (or actually even more inefficient!) than without a StagingArea.

The intended way of using it, is having a list of enqueue_ops. And you should pre-stage `n` samples. I don't think you're prestaging anything now, which means you don't have a buffer, which means you essentially don't really have any form of queue at all, and won't benefit.

Notice that often you want to use a cascade of queues (StagingAreas) and that means you'll need to do something like this:
```python
for i, op in enumerate(enqueue_ops):
    sess.run(enqueue_ops[:i])
```
The above loop will first enqueue on the first staging area, then the first+second, then the first+second+third, etc, unwill all your staging area cascade is filled. Then you can run the full `enqueue_ops` function `n` times, where `n` is your buffer size.

What complicates the above, is that if you want to empty your queues (and not use a continuous feed), then for the last few batches, you need to _stop+ using the enqueueing of (1) stop fetchign the first queue, (2) stop fetching the first+second queue (3) etc.

### Monkey patching of the backend
We could consider making a Keras MR that allows some kwargs with extra ops to be passed to the backend `Function` so that you don't need to monkey patch it. The way you're currently doing it (monkey patching) is indeed the only way.

Lastly, multi-gpu and multi-node won't help mnist of cifar probably. It doesn't matter at all, because those aren't real use cases anyway. In more practical use-cases (imagenet, automotive, etc) the models are deeper and the input data larger, and that's where your keras experiment will really shine. I'll see if I can help you there.

**Comment by ViaFerrata** [comment]
Don't want to change the subject, but

> In more practical use-cases (imagenet, automotive, etc) the models are deeper and the input data larger, and that's where your keras experiment will really shine.

Yes, indeed! I'm using the multi-gpu code with a WRN 16-4 and 3D data (13x18x50) and using 4 GTX1080 in a single node instead of one scales x4 :) Thanks a lot for your efforts @avolkov1

**Comment by avolkov1** [comment]
@TimZaman I see. Yea, I was not really sure what I was doing with the StagingArea. I put it there so that hopefully someone might fix it or tell me what I'm doing wrong. Thank you for that. I would greatly appreciate it if you helped me in getting that to work properly. I'm not sure how to implement the prestaging. There are very few examples and documentation is lacking.

With regards to mnist and cifar examples, I know those are silly and don't scale, but I'm using them purely for illustrating implementation details. Cifar scales a little bit and the reason I use Cifar10 is because the dataset for it is small and easy to download. Everyone is familiar with Cifar10.

I would really appreciate any help on porting this to imagenet with ResNet50 especially on our internal psgcluster (refer to this repo on yagr). I haven't had a chance to implement the ResNet50 benchmark which is what I really need to compare with native Tensorflow approach and perflab benchmarks. My goal is to have on par performance for ResNet50 with native Tensorflow by using Keras with tfrecords approach and this simple multigpu parallelization. I also want it to scale when I run distributed.

When I run Tensorflow with RDMA on psgcluster my current distributed implementation is not too bad given the very simple approach, but it can be much better I believe.

**Comment by avolkov1** [comment]
@ViaFerrata Awesome, I'm glad it's helping. Are you using fit_tfrecords approach, or fit with feed_dict?

**Comment by TimZaman** [comment]
@avolkov1 I'm currently running your code with imagenet+resnet18 (with tfrecords) and get 3x scaling speedup with 4x titan X. I'll see if the stagingarea's prequeueing can bump this up. 3x is already pretty fair.

**Comment by ViaFerrata** [comment]
@avolkov1 I'm using the Keras fit_generator function with hdf5 files, so feed_dict. Currently, I'm waiting for https://github.com/fchollet/keras/pull/6928, since the data feeding could be made (a lot?) faster with native framework input. Quite a shame that TF graph rewiring is needed for that :/

I think getting a super fast data pipe to work (TF-like performance) should be of high priority for Keras, especially for people that use multi-gpu or distributed training. Unfortunately, I can't really contribute to that.

**Comment by avolkov1** [comment]
@TimZaman Any progress. I would really like to see the resnet implementation and if you get the stagingarea working that would be awesome.

I will push an example today with distributed (on SLURM) training implementation where I achieved really good scaling and performance. 2 nodes each with 4 GPUs, reduced the training epoch to 2 sec./epoch on Cifar10 down from 12 sec./epoch for 1 GPU. I know, the Cifar10 example is a toy example, but I think the distribution will scale properly with more stressing examples.

**Comment by TimZaman** [comment]
Yeah I even got 3.6x scaling now [without queues/stagingarea even]. I hit a big snag with the stagingarea. I realized that both the label and the image need to be in the same queue (make sure things stay connected deterministically.
It makes no sense to have just a queue for one, and not for the other. That'll result is very nasty issues and data/label discrepancies.
How can we make sure we somehow enqueue the labels too? Sure, we can enqueue both at the same time _before_ we declare the tensors into `keras.layers.Input(tensor=..)`, but that would put the stagingarea/queue on the CPU. Although surely that should give a benefit, we want one per GPU, resident on the GPUs that'll use the data.
Not sure yet how to proceed there.

**Comment by ppwwyyxx** [comment]
@ViaFerrata if you are looking for fast data reading pipeline, [dataflow](http://tensorpack.readthedocs.io/en/latest/tutorial/dataflow.html) in tensorpack could be of interest. Although written in pure Python it can read ImageNet data fast enough to keep multigpu busy. A common misunderstanding is that native TFRecord make things faster -- not really, if your python side is already fast enough to fully utilize GPUs. 
I understand people may prefer keras-style APIs or hand-crafted training loops, that's why dataflow is written for use outside tensorpack as well. I noticed this repo actually contains some code I wrote there ... 😄

**Comment by avolkov1** [comment]
@ppwwyyxx Hey, yes, I took the dataflow portion out of tensorpack that you wrote. The repo here was meant for me to experiment with and share. I have an implementation using dataflow:
[https://github.com/avolkov1/keras_experiments/blob/master/examples/cifar/cifar10_cnn_mgpu_dflow.py](https://github.com/avolkov1/keras_experiments/blob/master/examples/cifar/cifar10_cnn_mgpu_dflow.py)

I did not see an immediate speedup. Looking at tensorpack code it was difficult to figure out how the code works for QueueInputTrainer, QueueInput, and EnqueueThread. I would love to understand it better. My implementation is here:
[https://github.com/avolkov1/keras_experiments/blob/master/keras_exp/mixin_models/dataflow.py](https://github.com/avolkov1/keras_experiments/blob/master/keras_exp/mixin_models/dataflow.py)

Any hints you might have I would appreciate it.

**Comment by TimZaman** [comment]
> A common misunderstanding is that native TFRecord make things faster

Correct: it's the tf-native queues (&stagingarea) that make it faster. But anyone that uses tfrecords often uses these queues - so there's a correlation.

@ppwwyyxx Depends on how much you care about speed. If you want good performance in TF, you need queues natively inside TF, and preferably queues that are already on the GPU. The latter constraint leaves you only with a StagingArea. The TF's Queue(Runners) to will also be deprecated in TF 2.0 because their perf is not satisfactory.
If you really care about that last +10~20% speed, a tf-native queue is a must. Then still you can use tensorpack, but that should then push into the queue natively, hence, you'll need tensors as inputs in Keras anyway.

**Comment by ppwwyyxx** [comment]
@avolkov1 The problem is, for dataset as small as cifar, the bottleneck is probably not in Python-side data loading -- you can load all of them in memory anyway. In my experience hiding the latency with queues will make more visible differences in the overall speed (probably not much as well, because cifar10 is really tooo small.... I'm wondering why everyone is talking about performance on such a tiny dataset..). But for dataset such as ImageNet, loading them in Python-side is already an issue, (and you'll also need queues and stagingareas). 

@TimZaman I understand the correction, that's why it's a misunderstanding. To be more precise, what I meant is tfrecord __itself__ doesn't make things faster.
I care about speed very much. We are on the same page that using queues/stagingareas is a must. But using native queues and StagingAreas doesn't mean I have to use tfrecord, as long as Python-side data loader is fast enough -- that's the point I wanted to make. tensorpack has included queues and StagingAreas already, btw.

Also I'm not saying dataflow alone will make training faster. To really avoid framework overhead, in general you need all the four blocks in parallel with training: (showing a figure from my documentation)
![0808-08 05 49](https://user-images.githubusercontent.com/1381301/29078895-7e1ce7e4-7c10-11e7-9bc0-1406c9f5fc85.png)

`tensorpack.dataflow` is a fast __Python loader__, i.e. it makes the first two blocks faster. Using tfrecord, you'll then be doing the first two blocks in TF and skipping the third block (which is probably fast as well but harder to write IMHO).

**Comment by bzamecnik** [comment]
@avolkov1:

> I would really appreciate any help on porting this to imagenet with ResNet50 

I tried to apply your multi-gpu code to Resnet50 and InceptionV3 from keras.applications with synthetic imagenet dataset:

https://github.com/rossumai/keras-multi-gpu/blob/master/keras_tf_multigpu/examples/benchmark_inception3_resnet50.py

On Azure 2xTesla M60:

For Resnet50 with PS=CPU 1.72x speed-up on 2 GPUs, with PS=GPU 1.65x speed-up, for InceptionV3 the numbers are almost the same.

**Comment by avolkov1** [comment]
@bzamecnik Thank you. Great stuff on your [blog](https://github.com/rossumai/keras-multi-gpu/blob/master/blog/docs/keras-tensorflow.md). I didn't know that Chollet added the GPU parallelization to the Keras utils. I don't know why he didn't fix the part with slicing on CPU. Maybe my repo here might become obsolete soon.
I'm going to move the enqueue/StagingArea stuff and monkey patching hack into a dev branch to clean up the code.

**Comment by bzamecnik** [comment]
Thanks. The code evolves so quickly... So far fchollet's version is just in master, not released yet. Yeah, I'm also missing the CPU slicing and also correct saving/loading. Maybe a space for another PR.

So far I'm trying to summarize the measurements. Then I'll try to measure new packages like horovod and tensorpack. But anyway we'd like to have a clean Keras code without other dependencies and with nice API.

So far the result is that double-buffering at GPU with StagingArea is the key. I'll try to make some experiments with that as well. It seem that it need only queue of size 2 for one GPU - one input batch being copied while another is being consumed by computation. The only serial part would then be only gradient and weight exchange.

**Comment by bzamecnik** [comment]
I tried to fiddling with StagingArea and Keras. To summarize: For at least double-buffering at GPU we need a queue of size 2. Ie. first we need to put(), then at each batch within model.fit() we need both put() and get() within the same session.run(), at the last batch we need to only get() or possibly put() a dummy value.

I tried to provide values for StagingArea via a placeholder fed with feed_dict (for simplicity). It turns out that in Keras it's possible to combine a tensor input and feed_dict input! Thus we can use tensors StagingArea.get() as main input and target_tensors labels, StagingArea.put() as additional operation to `fetches` argument and values for next batch as `x` argument to model.fit(). Note that features and labels are represented as a list of tensors and are transferred together in single get/put call, not separately.

Whole rudimentary code snippet: https://gist.github.com/bzamecnik/54104644baddc614230102e65f02b593

Problems to overcome:
- the API usage is very hacky, really not nice (the usage of pipelining should be encapsulated)
- we can call model.fit() on single batch (steps_per_epoch=1) since Keras doesn't automatically slice the batches in the tensor input mode
- since the session.run() is wired once in compile we should provide a dummy value for the last batch
- we need to switch between training and validation sets if validation set is provided to model.fit()
- it needs patch of K.function() (I rewrote the monkey patch to a Keras fork that will go in a PR): https://github.com/bzamecnik/keras/commit/8593309c371ce716fd039e33ed5ae4079096ee0f

TF queues API might help in cleaner API - we only add a few more ops to fetches and run Python->TF transfer in another thread. But it's deprecated. Newer Dataset API in TF 1.3 doesn't support Python memory as data source, except for baking numpy arrays as constant (no-go). In TF 1.4 we have in addition Dataset.from_generator() which seems to be a viable alternative, but TF 1.4 is not out yet.

**Comment by TimZaman** [comment]
tf 1.4 is already in rc.

If you want to use the `StagingArea` just:

1. make sure your model input is from the stagingarea's `.get()`
2. when you have only 1 stagingarea, manually `.put()` 1 element in there before `fit`ting
2. for each step, add a `.put()` to keras's `sess.run` list deep down in the backend

**Comment by avolkov1** [comment]
@bzamecnik  @TimZaman 
If you can illustrate with an example and/or PR how I could enable this for multigpu via my mini-framework here I would really appreciate that. Thank you.

**Comment by bzamecnik** [comment]
Yeah, that's my goal. However, first I have to finish a working mechanism for 1 GPU.

**Comment by TimZaman** [comment]
If you're interested in multi-gpu or multi-node with keras, just use horovod, it has the best scaling by far: https://github.com/uber/horovod

**Comment by bzamecnik** [comment]
Another proof-of-concept step: https://gist.github.com/bzamecnik/3c2b5279a5949d694421d7cfbe813557

Inputs to StagingArea provided via Variable assignment which can be decoupled from Keras' Session.run(). The pipelining logic is encapsulated in a callback. It's able to slice an input numpy array to batches.

Still it relies on feed_dict, but we don't have to hack with Keras inputs. It still needs to take care of validation set/prediction etc. And it also assumes one feature input and one label output.

**Comment by bzamecnik** [comment]
@TimZaman Yeah, I'd like to try. So far I didn't find time for that. Their results look impressive. Still solving StagingArea could improve even single-GPU performance in Keras.

**Comment by avolkov1** [comment]
I installed and played around with horovod. I like their synchronous optimizer and their simplification of Tensorflow API is nice, but if one can get away without synchronizing/averaging gradients, then it's not speeding up Keras training. Here are the results I got for 2 GPUs mnist example that they posted (I split the data set of 60,000 samples into two data sets of 30,000 samples processed on each GPU):
```
https://github.com/uber/horovod/blob/master/examples/keras_mnist.py
# my code modification
nsamples = x_train.shape[0]
mysamples = nsamples // hvd.size()
start_sam = hvd.local_rank() * mysamples
stop_sam = min((hvd.local_rank() + 1) * mysamples, nsamples)
x_train = x_train[start_sam:stop_sam, ...]
y_train = y_train[start_sam:stop_sam, ...]
```

```
# mpirun -np 2 python keras_mnist.py
Train on 30000 samples, validate on 10000 samples
Train on 30000 samples, validate on 10000 samples
Epoch 1/6
30000/30000 [==============================] - 10s - loss: 0.3384 - acc: 0.8965 - val_loss: 0.0729 - val_acc: 0.9769
Epoch 1/6
30000/30000 [==============================] - 10s - loss: 0.3381 - acc: 0.8974 - val_loss: 0.0729 - val_acc: 0.9769
Epoch 2/6
30000/30000 [==============================] - 6s - loss: 0.1007 - acc: 0.9697 - val_loss: 0.0468 - val_acc: 0.9834
Epoch 2/6
30000/30000 [==============================] - 6s - loss: 0.1039 - acc: 0.9690 - val_loss: 0.0468 - val_acc: 0.9834
Epoch 3/6
30000/30000 [==============================] - 6s - loss: 0.0747 - acc: 0.9787 - val_loss: 0.0372 - val_acc: 0.9876
Epoch 3/6
30000/30000 [==============================] - 6s - loss: 0.0749 - acc: 0.9769 - val_loss: 0.0372 - val_acc: 0.9876
Epoch 4/6
30000/30000 [==============================] - 6s - loss: 0.0604 - acc: 0.9816 - val_loss: 0.0318 - val_acc: 0.9889
Epoch 4/6
30000/30000 [==============================] - 6s - loss: 0.0672 - acc: 0.9801 - val_loss: 0.0318 - val_acc: 0.9889
Epoch 5/6
30000/30000 [==============================] - 6s - loss: 0.0504 - acc: 0.9853 - val_loss: 0.0363 - val_acc: 0.9872
Epoch 4/6
30000/30000 [==============================] - 6s - loss: 0.0549 - acc: 0.9829 - val_loss: 0.0363 - val_acc: 0.9872
Epoch 6/6
30000/30000 [==============================] - 6s - loss: 0.0434 - acc: 0.9860 - val_loss: 0.0308 - val_acc: 0.9897
Epoch 6/6
30000/30000 [==============================] - 6s - loss: 0.0478 - acc: 0.9853 - val_loss: 0.0308 - val_acc: 0.9897
Test loss: 0.0308260866315
Test accuracy: 0.9897
Test loss: 0.0308260866315
Test accuracy: 0.9897
```
Compare to [`mnist_cnn.py`](https://github.com/fchollet/keras/blob/master/examples/mnist_cnn.py) that I prallelized on 2 GPUs using asynchronous training:
```
# CUDA_VISIBLE_DEVICES=0,1 python mnist_cnn_mgpu.py
Train on 60000 samples, validate on 10000 samples
Epoch 1/6
60000/60000 [==============================] - 7s - loss: 0.3254 - acc: 0.9017 - val_loss: 0.0725 - val_acc: 0.9764
Epoch 2/6
60000/60000 [==============================] - 3s - loss: 0.1113 - acc: 0.9672 - val_loss: 0.0509 - val_acc: 0.9827
Epoch 3/6
60000/60000 [==============================] - 3s - loss: 0.0826 - acc: 0.9753 - val_loss: 0.0419 - val_acc: 0.9852
Epoch 4/6
60000/60000 [==============================] - 3s - loss: 0.0712 - acc: 0.9784 - val_loss: 0.0365 - val_acc: 0.9874
Epoch 5/6
60000/60000 [==============================] - 3s - loss: 0.0622 - acc: 0.9812 - val_loss: 0.0337 - val_acc: 0.9884
Epoch 6/6
60000/60000 [==============================] - 4s - loss: 0.0544 - acc: 0.9831 - val_loss: 0.0326 - val_acc: 0.9887
Test loss: 0.0326202042509
Test accuracy: 0.9887
```
For single GPU each epoch runs in ~5 sec. on P40s that I am testing on. Asynchronous training doesn't work in all cases, so in those cases horovod should be very valuable.

I'm curious to compile Tensorflow with NCCL2 and perform gradient synchronization without MPI hooks. For distributed I get pretty good results using GRPC with RDMA (`protocol='grpc+verbs'` and trying to get `protocol='grpc+gdr' in tf1.4 working which should improve distributed training even more).

**Comment by TimZaman** [comment]
I don’t see a real-world usecase for multigpu/node for networks as shallow
as LeNet and MNIST datasets. Try resnet18 and up/inception/vgg and you will
get amazing numbers. 7x scaling with 8 GPUs is the last I’ve seen with
resnet50 equivalent complexity

**Comment by avolkov1** [comment]
I'll test horovod with resnet (trust but verify). Their metrics are based on images/sec., but each epoch is actually slower. If the walltime to convergence is faster, then that's what matters. In more complicated networks such as the variational autoencoder examples, LSTM/RNNs (ptb lstm example based on Keras), the asynchronous training sped up the training, but did not work as far as convergence. I briefly tested horovod on those examples and it worked, but did not speed up the training (single GPU performance was faster and achieved the same convergence rate or even faster).

The asynchronous training I have implemented speeds up epochs or iterations, but I have observed that the async approach sometimes messes with loss and accuracy metrics, and converges slowly or doesn't converge at all. Hyperparameter adjustments with async training are more complicated. I also tried implementing the synchronous gradient averaging ([`OptimizerMultiGPUMixin`](https://github.com/avolkov1/keras_experiments/blob/master/keras_exp/multigpu/optimizers.py#L63)), but I don't think it's working correctly (Note: sync doesn't seem to work right even with the fix to issue #12 for compatibility with Keras 2.0.8). I verified that the horovod gradient averaging is definitely working as opposed to my attempt at it.

**Comment by bzamecnik** [comment]
Hi, I made some experiments to figure out behavior of StagingArea. The results are cheering :)

- [it supports variable batch size, while enforcing data sample shape](https://gist.github.com/bzamecnik/05ff3bda1f82b16e325a3a1857fd2994)
- [it's possible to to pipelining and join get() and put() into a single operation with tf.tuple()](https://gist.github.com/bzamecnik/a57487723b95963fb73921c91b9950ae)
  - this removed the need for patching K.function() since put() operation is hidden in the input tensor
  - but I'm not sure whether this would allow memcpy parallel to computation, most probably not

**Comment by bzamecnik** [comment]
Also I tried to run make_parallel on 4xM60 (on Azure NV24) and got nice results:

|speedup|inception3|cifar10 convnet|
|-|-|-|
|tf_cnn_benchmarks|3.86x|3.54x|
|Keras + avolkov1|3.04x|2.57x|

- speedup on 4x M60 wrt. 1x M60
- cifar10 with batch size 512
- inception3 with batch size 32
- PS=CPU

Speedup for Keras is significant, but still there's room for improvement.

**Comment by ppwwyyxx** [comment]
There is no need to use tf.tuple with control_inputs, which adds an unnecessary op dependency: 
> control_inputs contains additional ops that have to finish before this op finishes, but whose outputs are not returned.

You can just `tf.group` the get and put op.

**Comment by bzamecnik** [comment]
@ppwwyyxx Yeah, true, that I don't want that dependency. But unfortunately, [tf.group](https://www.tensorflow.org/api_docs/python/tf/group) doesn't return any value (in this case from get()): "This op has no output.". I checked this behavior.

**Comment by ppwwyyxx** [comment]
You can still use `get` for the training, and use `tf.group` to group `get`, `put` as well as the training op, just to make sure `get` and `put` will be run, without introducing dependencies.
But then I guess you can't avoid patching Keras, but I'm not familiar with that so can't comment much.

**Comment by bzamecnik** [comment]
@ppwwyyxx Unfortunately we can't get value from `StaginArea.get()` through `tf.group()`. Check this example: https://gist.github.com/bzamecnik/44a3642c05afc2ac09135687edce8a9d Do you see any flaw in that or you have a different mechanism in mind? Thanks.

Yeah, probably the way to go to enable parallel put() would be using the patch of K.function().

**Comment by ppwwyyxx** [comment]
I know `group()` returns no value. You can use the output of `t = get()` for training, and use `op = tf.group(t, put())` just for dependencies. `get()` will not be evaluated twice. But then `op` may need to be patched into Keras.

**Comment by bzamecnik** [comment]
Ah, yes, I understand. If we just provide `t` to the `Input` layer that `tf.group` including `put()` will not get executed since it is not linked to `fetches` in `Session.run()`. That's why now we provide `put()` via `enqueue_ops` to be included there. I was interested if it's possible to avoid that by chaining `put()` in the tensor that goes to `Input`.

**Comment by bzamecnik** [comment]
Hooray. I managed to get the pipelining to work and observe the desired effect on a basic CIFAR10 convnet. :)

https://gist.github.com/bzamecnik/b9dbd50cdc195d54513cd2f9dfb7e21b

<img width="1268" alt="screenshot 2017-10-28 v 21 29 45" src="https://user-images.githubusercontent.com/446124/32137825-9cba6cd8-bc27-11e7-9cb5-58ce727ea04d.png">

It works well with variable-sized batches. Also the throughput on a single GPU increased a tiny bit.

So far still the data is fed into the StagingArea in a separate session.run() via an intermediate Variable. This is responsible for the initial gap. It can be reduced by TF queue or possibly Dataset 
API (?).

A known limitation for tensor inputs to Keras Model is that they do not support validation data yet. We can still perform validation the classic way via feed_dict.

**Comment by TimZaman** [comment]
The used models and datasets make little sense to me work benchmarking;
they are too small and narrow and therefore barely reveal meaningful
numbers and bottlenecks.

I don’t see how the stagingarea’s put command is used in the snippet in a
way that would speed things up, you are calling it synchronously it seems;
whereas you should be calling it simultaneously with your get command in
the same session.run fetch statement to have both put and get run
independently and asynchronously.

**Comment by bzamecnik** [comment]
@TimZaman Of course I'd like to try it also on Inception3/Resnet50 on imagenet-synth, but so far adapting a smaller CIFAR10 convnet was easier. Indeed get() and put() are called in the same session.run(): tensor from get() goes to Input layer and put() is passed in model.compile() to fetches. Copy from Python to TF memory is still synchronous (it was not the point of this experiment) and nvprof shows that the host-to-device memcpy is really asynchronous. The gap between computations was reduced (in this case roughly from 7 to 4 ms).

**Comment by bzamecnik** [comment]
In order to make experiments with the StagingArea pipelining easier I extracted the code into [StagingAreaCallback](https://github.com/rossumai/keras-multi-gpu/blob/master/keras_tf_multigpu/callbacks.py#L10) and added [a few examples](https://github.com/rossumai/keras-multi-gpu/tree/master/keras_tf_multigpu/examples/bzamecnik/staging_area) - trivial MNIST, simple CIFAR10 and ResNet50.

The next step would be to try to integrate it with multi-GPU code and asynchronous Python->TF feeding (TF queue/Dataset).

**Comment by TimZaman** [comment]
> A known limitation for tensor inputs to Keras Model is that they do not support validation data yet. We can still perform validation the classic way via feed_dict.

It is supported: I added that myself to Keras months ago: https://github.com/fchollet/keras/commit/135efd66d00166ca6df32f218d086f98cc300f1e#diff-299cfd5886683a4b012f286403769fc1

The problem you're solving here isn't that big, and your solution won't give great results, if any at all, as your numpy code is still in the critical path. It would make sense if you would use an example that natively uses tensorflow tensors (f.e. tfrecords), which would be the only logical usecase of the stagingarea; unless maybe your numpy code is multiprocess/threaded. Furthermore, try to place the stagingarea on the gpu for extra juice. I'll be unsubscribing from this thread as there are way too many messages posted here.

**Comment by bzamecnik** [comment]
Thanks. Sorry, for too much noise. I was pointing at `validation_data` in fit(), not `evaluate()` for testing. The StagingArea is located on GPU, otherwise it makes no sense.

**Comment by TimZaman** [comment]
Why wouldn't the stagingarea on the CPU make no sense? In typical high-performance jobs you've got a tensorflow-native graph, meaning there's no feed_dicts and no slow numpy stuff going on.

Anyway, I'm adding support for everyone to add their stagingArea to upstream keras with this MR: https://github.com/fchollet/keras/pull/8285 . There's some pseudocode in that MR of how to use the StagingArea.

**Comment by avolkov1** [comment]
Thank you guys very much. I see you both created PRs fchollet/keras#8285 and fchollet/keras#8286. Could you guys come to an agreement and combine the PRs into one PR? But regardless, I'd like to apply the PRs in my monkey patch temporary code here until it gets incorporated (in some fashion) into mainstream Keras.
Thank you both again. Very much appreciated!

**Comment by bzamecnik** [comment]
And our article *Towards Efficient Multi-GPU Training in Keras with TensorFlow* about multi-GPU training on Keras+TensorFlow has been published: https://twitter.com/RossumAi/status/924985495851012096

**Comment by TimZaman** [comment]
I benchmark a ton; and multi-gpu training for TF backends doesn't get much better than just using Horovod for it.

**Comment by avolkov1** [comment]
@TimZaman Yes indeed! I've been testing horovod over the weekend. For some reason on my install on psgcluster I can't get horovod to run well. Is there some trick in installing horovod without containers and getting it to work well? I compiled tf1.4 (whatever was current two weeks ago for tf1.4) with CUDA9, and installed horovod with NCCL2, but something must have been off.

I installed horovod in our dgx containers, and indeed, it scaled even for the trivial mnist example. I'll add horovod examples to my repo. We don't have any dockerfiles or containers available publicly with NCCL2 yet, so it will be kind of difficult for other to replicate. I repeated my test that I ran before and this is what I'm getting on P40s:
```
        GPU0    GPU1    GPU2    GPU3    mlx5_0  CPU Affinity
GPU0     X      SOC     SOC     SOC     SOC     0-15
GPU1    SOC      X      PHB     PHB     PHB     16-31
GPU2    SOC     PHB      X      PIX     PHB     16-31
GPU3    SOC     PHB     PIX      X      PHB     16-31
mlx5_0  SOC     PHB     PHB     PHB      X

# CUDA_VISIBLE_DEVICES=1,2 mpirun --report-bindings -np 2 -H localhost  -rf rankfile_hswp100  python mnist_horovod.py
Epoch 1/12
30000/30000 [==============================] - 3s - loss: 0.3377 - acc: 0.8963 - val_loss: 0.0713 - val_acc: 0.9763
Epoch 1/12
30000/30000 [==============================] - 3s - loss: 0.3330 - acc: 0.8957 - val_loss: 0.0713 - val_acc: 0.9763
Epoch 2/12
30000/30000 [==============================] - 2s - loss: 0.1048 - acc: 0.9690 - val_loss: 0.0478 - val_acc: 0.9843
Epoch 2/12
30000/30000 [==============================] - 2s - loss: 0.1069 - acc: 0.9688 - val_loss: 0.0478 - val_acc: 0.9843
...
...
...
Epoch 10/12
30000/30000 [==============================] - 2s - loss: 0.0315 - acc: 0.9904 - val_loss: 0.0254 - val_acc: 0.9910
Epoch 10/12
30000/30000 [==============================] - 2s - loss: 0.0316 - acc: 0.9899 - val_loss: 0.0254 - val_acc: 0.9910
Epoch 11/12
30000/30000 [==============================] - 2s - loss: 0.0282 - acc: 0.9910 - val_loss: 0.0267 - val_acc: 0.9910
Epoch 11/12
30000/30000 [==============================] - 2s - loss: 0.0294 - acc: 0.9901 - val_loss: 0.0267 - val_acc: 0.9910
Epoch 12/12
30000/30000 [==============================] - 2s - loss: 0.0246 - acc: 0.9922 - val_loss: 0.0285 - val_acc: 0.9908
Epoch 12/12
30000/30000 [==============================] - 2s - loss: 0.0285 - acc: 0.9909 - val_loss: 0.0285 - val_acc: 0.9908
Test loss: 0.0285155087762
Test accuracy: 0.9908
Test loss: 0.0285155087762
Test accuracy: 0.9908
```
The above for single GPU runs at 4s per epoch, and for 4 GPUs 1s. It seems to scale linearly at least within a single node (I haven't been able to get it to work distributed yet).

I've tested it with more complicated keras (ptb lstm, variation autoencoder, ...) networks and it generalizes pretty well. I've had issues running the asynchronous training approach with such networks (loss/acc diverge for multi-GPUs with asynchronous training), but horovod worked.

**Comment by ViaFerrata** [comment]
@avolkov1 That would be great!
Can you use it directly with Keras objects like they describe it beneath 'Usage'?
E.g. a Keras optimizer in `hvd.DistributedOptimizer` ? I'm also a bit clueless how to handle "5." in e.g. fit_generator.

Also, as far as I get it, the gradients in Horovod are averaged synchronously? And in your (great!) code it is asynchronous?

**Comment by TimZaman** [comment]
Aleks, come on, stop using MNIST for distributed training.
Are you in Endeavor any time soon, to meet?

**Comment by avolkov1** [comment]
Sorry about resorting to mnist and cifar, but these datasets are universally available.

@ViaFerrata Yea, I'll post some code. I'm not sure about fit_generator. I'll be traveling for the next day, but once I get a chance I'll add some examples. Do this:
```
hvd.init()  # globally up top is a good idea

def main():  # in main do
    # Pin GPU to be used to process local rank (one GPU per process)
    config = tf.ConfigProto()
    config.gpu_options.allow_growth = True
    config.gpu_options.visible_device_list = str(hvd.local_rank())
    K.set_session(tf.Session(config=config))

    # define some optimizer
    opt = tf.train.RMSPropOptimizer()
    # Add Horovod Distributed Optimizer.
    opt = hvd.DistributedOptimizer(opt)  # , use_locking=True)

    # run prior to fit
    sess.run(hvd.broadcast_global_variables(0))
```
And don't miss any steps, because if you do it might run, but results will be messed up. Yes, horovod implements synchronous training whereby it averages gradients.
Maybe once I understand the code well enough I could add a proper synchronous gradient mode to my code base without resorting to MPI.
The code is a mess, but if you want to see a more involved example I've been working with I attached it.

[variational_autoencoder_deconv_tfqueue_horovod.txt](https://github.com/avolkov1/keras_experiments/files/1428939/variational_autoencoder_deconv_tfqueue_horovod.txt)

There I'm using tf-queue, horovod, keras, etc. For mnist it runs faster but doesn't really converge faster because it converges fast enough with 1 GPU, but for larger datasets (with large images), it should run a lot faster when scaling to multi-gpus.

Here's [mnist_horovod.txt](https://github.com/avolkov1/keras_experiments/files/1428954/mnist_horovod.txt) a basic mnist example I benchmarked. I'll modify it with tfqueue. It runs a little bit faster with tfqueue, but with tfqueue I don't have to split the data I just reduce steps per epoch so the random sampling is done correctly. I do this because I prefer to decrease the epoch walltime time instead of increasing imgs/sec metric. With img/sec increase overall training time is reduced the same I suppose, as long as convergence is faster with less epochs, but I prefer to reduce epoch iteration time and "split" data across GPUs. This makes it clear that 12 epochs multigpu is faster than 12 epochs 1 gpu.

**Comment by ppwwyyxx** [comment]
I tested uber's benchmark script (https://github.com/tensorflow/benchmarks/pull/73) on a single DGX1. For resnet50 horovod is about 95% the speed of native tensorflow (replicated mode), which is really good considering it's in multi-process.
//UPDATE(2018): a later benchmark shows that horovod has slightly better performance
